zkir/Dialect/Field/Conversions/FieldToModArith/Extension/QuarticExtensionField.cpp

```diff
@@ -0,0 +1,534 @@
+  Value bDiff2 = b.create<mod_arith::SubOp>(b0, b1x2);
+  bDiff2 = b.create<mod_arith::AddOp>(bDiff2, b2x4);
+  bDiff2 = b.create<mod_arith::SubOp>(bDiff2, b3x8);
+  auto v4 = b.create<mod_arith::MulOp>(aDiff2, bDiff2);
+
+  // v5 = (a₀ + 3a₁ + 9a₂ + 27a₃)(b₀ + 3b₁ + 9b₂ + 27b₃)
+  Value aSum3 = b.create<mod_arith::AddOp>(a0, a1x3);
+  aSum3 = b.create<mod_arith::AddOp>(aSum3, a2x9);
+  aSum3 = b.create<mod_arith::AddOp>(aSum3, a3x27);
+  Value bSum3 = b.create<mod_arith::AddOp>(b0, b1x3);
+  bSum3 = b.create<mod_arith::AddOp>(bSum3, b2x9);
+  bSum3 = b.create<mod_arith::AddOp>(bSum3, b3x27);
+  auto v5 = b.create<mod_arith::MulOp>(aSum3, bSum3);
+
+  // v6 = a₃ * b₃
+  auto v6 = b.create<mod_arith::MulOp>(a3, b3);
+
+  return toomCookInterpolate(v0, v1, v2, v3, v4, v5, v6);
+}
+
+Value QuarticExtensionField::frobeniusMap(Value x, unsigned exponent) {
+  // Frobenius map: φ^e(x₀ + x₁v + x₂v² + x₃v³)
+  // Using precomputed coefficients.
+  // v^(k*p^e) = ξ^(k*p^e / 4) * v^(k*p^e mod 4)
+
+  if (exponent == 0) {
+    return x;
+  }
+
+  auto coeffs = toCoeffs(b, x);
+  unsigned idx = exponent - 1; // frobeniusCoeffs is 0-indexed for e=1,2,3
+
+  // Initialize result coefficients to precomputed zero
+  SmallVector<Value, 4> result = {zero, zero, zero, zero};
+
+  // For each input coefficient k, add scaled value to appropriate output
+  // position
+  for (unsigned k = 0; k < 4; ++k) {
+    unsigned newPos = frobeniusPowers[idx][k];
+    auto scaled =
+        b.create<mod_arith::MulOp>(coeffs[k], frobeniusCoeffs[idx][k]);
+    result[newPos] = b.create<mod_arith::AddOp>(result[newPos], scaled);
+  }
+
+  return fromCoeffs(b, type, result);
+}
+
+Value QuarticExtensionField::inverse(Value x) {
+  // Frobenius-based inverse for quartic extension fields.
+  // See Tachyon's quartic_extension_field.h for reference.
+  // Algorithm 11.3.4 in "Handbook of Elliptic and Hyperelliptic Curve
+  // Cryptography"
+  //
+  // For Fp4, we use the formula:
+  //   a⁻¹ = a^(r-1) * (a^r)^(-1)
+  // where r = p³ + p² + p + 1 = (p⁴ - 1) / (p - 1)
+  //
+  // a^r = a * φ(a) * φ²(a) * φ³(a) is the norm and lies in Fp.
+  // a^(r-1) = φ(a) * φ²(a) * φ³(a)
+  //
+  // Important: We compute (a^r).c0 directly using the formula:
+  //   (a^r).c0 = a.c0 * b.c0 + ξ * (a.c1*b.c3 + a.c2*b.c2 + a.c3*b.c1)
+  // where b = a^(r-1). This avoids relying on Frobenius fixing only Fp.
+
+  // φ(a)
+  auto phi1 = frobeniusMap(x, 1);
+
+  // φ²(a)
+  auto phi2 = frobeniusMap(x, 2);
+
+  // φ³(a)
+  auto phi3 = frobeniusMap(x, 3);
+
+  // a^(r-1) = φ(a) * φ²(a) * φ³(a)
+  auto aRMinus1 = mul(phi1, phi2);
+  aRMinus1 = mul(aRMinus1, phi3);
+
+  // Compute (a^r).c0 directly without full multiplication.
+  // For a * b where a = (a0, a1, a2, a3), b = (b0, b1, b2, b3):
+  //   c0 = a0*b0 + ξ*(a1*b3 + a2*b2 + a3*b1)
+  auto aCoeffs = toCoeffs(b, x);
+  auto bCoeffs = toCoeffs(b, aRMinus1);
+
+  // a0*b0
+  auto a0b0 = b.create<mod_arith::MulOp>(aCoeffs[0], bCoeffs[0]);
```

Comment:
Please follow the naming convention described https://github.com/fractalyze/zkir/pull/114/commits/e57672df908c4ff2b68d4f61d20c5a39b92b4de4